File: LeetCode - Algorithm/Binary Tree/437.js
// 437. Path Sum III
// https://leetcode.com/problems/path-sum-iii/

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number}
 */
var pathSum = function (root, targetSum) {
  let result = 0;

  const getPathCounts = (node, currentSums = [0]) => {
    if (!node) return;

    currentSums.forEach((current) => {
      if (current + node.val === targetSum) result += 1;
    });

    const newCurrentSums = currentSums.map((sum) => sum + node.val);
    newCurrentSums.push(0);

    getPathCounts(node.left, newCurrentSums);
    getPathCounts(node.right, newCurrentSums);
  };

  getPathCounts(root);

  return result;
};

// [[3]]
// [[-2]]
// [[3, 3], [-2, 3]]

`
       10
    5     -3
  3   2     11
3 -2    1
`;

`
10, [0]
5 [10, 0]
3 [15, 5, 0]

-3 [10, 0]
11 [7, -3]
`;

File: LeetCode - Algorithm/Binary Tree/437.java
// 437. Path Sum III
// https://leetcode.com/problems/path-sum-iii/

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int pathSum(TreeNode root, int targetSum) {
        if (root == null)
            return 0;

        return this.calculatePathSum(root, targetSum, false);
    }

    public int calculatePathSum(TreeNode node, long targetSum, boolean isAdding) {
        int count = 0;

        if (targetSum == node.val)
            count += 1;

        if (node.left != null) {
            if (!isAdding)
                count += this.calculatePathSum(node.left, targetSum, false);

            count += this.calculatePathSum(node.left, targetSum - node.val, true);
        }

        if (node.right != null) {
            if (!isAdding)
                count += this.calculatePathSum(node.right, targetSum, false);

            count += this.calculatePathSum(node.right, targetSum - node.val, true);
        }

        return count;
    }
}

/*
 * In Java, the 'int' type can store values ranging from -2,147,483,648 to
 * 2,147,483,647 (Roughly -2 × 10^9 to 2 × 10^9).
 * 
 * In this problem, the constraints for a node's value is -10^9 <= Node.val <=
 * 10^9, meaning that adding or subtracting just two or three such values can
 * exceed the 'int' range. Therefore, I need to use the 'long' type for this
 * problem. I must be cautious when performing arithmetic operations in Java and
 * check the constraints.
 */